### PR TITLE
fix(rome_js_analyzer, rome_js_parser): suppress false positive diagnostics related to index signature parameter

### LINKED ISSUE #4634: 🐛 correctness/noUnusedVariables false positive when using indexed type
### Environment information

```block
playground
```

### What happened?

https://docs.rome.tools/playground/?lintRules=all&code=ZQB4AHAAbwByAHQAIABmAHUAbgBjAHQAaQBvAG4AIABmAGEAaQBsAHMAKABuAGEAbQBlADoAIABzAHQAcgBpAG4AZwApACAAewAKACAAIAByAGUAdAB1AHIAbgAgAHsACgAgACAAIAAgAG4AYQBtAGUALAAKACAAIAB9ACAAYQBzACAAewAgAFsAbgBhAG0AZQA6ACAAcwB0AHIAaQBuAGcAXQA6ACAAcwB0AHIAaQBuAGcAIAB9ADsACgB9AAoACgBlAHgAcABvAHIAdAAgAGYAdQBuAGMAdABpAG8AbgAgAHAAYQBzAHMAKABuAGEAbQBlADoAIABzAHQAcgBpAG4AZwApACAAewAKACAAIAByAGUAdAB1AHIAbgAgAHsACgAgACAAIAAgAG4AYQBtAGUALAAKACAAIAB9ACAAYQBzACAAewAgAFsAbgBvAHQATgBhAG0AZQA6ACAAcwB0AHIAaQBuAGcAXQA6ACAAcwB0AHIAaQBuAGcAIAB9ADsACgB9AAoACgAKAAoA

image

### Expected result

should not error

### Code of Conduct

- [X] I agree to follow Rome's Code of Conduct

### PR DESCRIPTION
**Pull request by nissy-dev**
## Summary

Fix #4634

The reason of this issue is that semantic analyzer handles index signature parameter as JsIdentifierBinding.
Thus, I add a new CST Node, TsIndexSignatureIdentifierBinding and suppress false positive diagnostics.

## Test Plan

I update some snapshot tests.

### CODE REVIEW

File: crates/rome_js_analyze/src/semantic_analyzers/suspicious/no_redeclare.rs [hunk 1]
```diff
@@ -128,72 +126,19 @@ fn check_redeclarations_in_single_scope(scope: &Scope, redeclarations: &mut Vec<
...
-                            name,
-                            declaration: *first_text_range,
-                            redeclaration: id_binding.syntax().text_trimmed_range(),
-                        }),
-                    }
+                    redeclarations.push(Redeclaration {
+                        name,
+                        declaration: *first_text_range,
+                        redeclaration: id_binding.syntax().text_trimmed_range(),
+                    })
```
Comment (nissy-dev) [inline crates/rome_js_analyze/src/semantic_analyzers/suspicious/no_redeclare.rs:137]: This rule no longer handles the case like interface A { [index: number]: string; [index: number]: string; }, so I removed unnecessary logic.

File: crates/rome_js_analyze/tests/specs/suspicious/noRedeclare/invalid.jsonc [hunk 2]
```diff
@@ -24,7 +24,5 @@
 	"function f() { var a; var a; }",
 	"function f() { var a; if (test) { var a; } }",
 	"for (var a, a;;);",
-	"for (;;){ var a, a,;}",
-	"function f(x) { var x = 5; }",
```
Comment (nissy-dev) [inline crates/rome_js_analyze/tests/specs/suspicious/noRedeclare/invalid.jsonc:28]: This case was not correctly checked, so I removed. (see: #4291)

File: crates/rome_js_analyze/tests/specs/suspicious/noRedeclare/invalid.jsonc [hunk 3]
```diff
@@ -24,7 +24,5 @@
...
-  "interface A { [index: number]: string; [index: number]: string; }"
```
Comment (nissy-dev) [inline crates/rome_js_analyze/tests/specs/suspicious/noRedeclare/invalid.jsonc:29]: semantic analyzer no longer handle this case, so I removed.

**Comment by ematipico** [comment]
@nissy-dev have you looked at the regressions?

**Review by Conaclos — APPROVED** [review]

**Comment by nissy-dev** [comment]
@ematipico I'll check it in a few days! But, I have never seen these `.symbol` files, so I would appreciate if you could tell me some information about symbols/microsoft tests.

**Comment by ematipico** [comment]
> @ematipico I'll check it in a few days! But, I have never seen these `.symbol` files, so I would appreciate if you could tell me some information about symbols/microsoft tests.

No worries! In Rome we have some conformance tests that run against some code bases, and one of them in the TypeScript repository. We do that using [git submodules](https://github.com/rome/tools/blob/main/CONTRIBUTING.md#cargo-coverage).

Our test suite runs against TypeScript conformance tests, and we check if our parser and their parser result match. If our parser doesn't match TS result, it's a fail.

In your case, it seems that what we could parse, now doesn't parse anymore as expected. The files listed in the comment come from the TypeScript repository, here's an example:
- https://github.com/microsoft/TypeScript/blob/main/tests/baselines/reference/computedPropertyNamesContextualType5_ES6.symbols
- https://github.com/microsoft/TypeScript/blob/main/tests/baselines/reference/parserSymbolIndexer4.symbols

Our test suite parsers the symbol file to collect the [correct data](https://github.com/rome/tools/blob/main/xtask/coverage/src/symbols/msts.rs#L213-L221), and [eventually we run it using our `xtask`](https://github.com/rome/tools/blob/main/xtask/coverage/src/runner.rs#L220).

Now, I don't precisely know the assertions that we do against our parser and what caused the regression, unfortunately. Although, a regression is something that we should not overlook.

**Something to consider**: the TS parser identifies those as [normal identifiers](https://ts-ast-viewer.com/#code/G4QwTgBAHgXBDeAoCKIG0DOcME8C2ARgPYA2AutgC5gCWAdgOYDciAvkA). Try to hover `x` and `s`, and both are `Identifier`. Maybe that's the reason why we used the same node. _Perhaps_, we could change the semantic model to ignore those identifiers if they are inside an interface/type.

**Comment by nissy-dev** [comment]
@ematipico Thank you for your thoughtful comments!

After checking these regressions, the reason is that the conformance test depends on the semantic analyzer. (see:  https://github.com/rome/tools/blob/54b2ab7fc456f4f52de97806ba21831d3c2f73e8/xtask/coverage/src/symbols/msts.rs#L83) Considering this point, the regressions are reasonable, so we can ignore.

> Perhaps, we could change the semantic model to ignore those identifiers if they are inside an interface/type.

Because of the reason, I confirmed the regressions will also happens even if I fix it as you pointed out.

**Comment by ematipico** [comment]
I believe it's fine if you decide to create a new node, but I think that the semantic analyzer should be updated to track the new node as a **symbol**.

It's fine if we fix the rules, but I think it's not fine if we regress the conformance without a specific reason.

**Comment by nissy-dev** [comment]
> I think that the semantic analyzer should be updated to track the new node as a symbol.

The symbol collected in this conformance test is some kinds of the semantic events.

Implementation: https://github.com/rome/tools/blob/54b2ab7fc456f4f52de97806ba21831d3c2f73e8/xtask/coverage/src/symbols/msts.rs#L83-L101

In my PR, the semantic analyzer no longer push the semantic event when entering an identifier binding node of index signatures. Your suggestion is that it would be better to solve this issue while handling an identifier binding of index signatures as the semantic event?

I am not sure if this is possible yet.  I have felt that many of the semantic analyzer functions don't take into account the identifier of index signatures, so a lot of changes may be needed.

**Comment by ematipico** [comment]
> I am not sure if this is possible yet. I have felt that many of the semantic analyzer functions don't take into account the identifier of index signatures, so a lot of changes may be needed.

I know. Although, I think this is not the correct fix. For example, the `usNamingConvetion` rule had a regression where those bindings aren't checked anymore. 

So, as long as we want to add a new node, we should still keep tracking in the semantic analyzer. Then, inside the `noDeclare` rule, we ignore all those bindings that belong to this new node.

**Comment by nissy-dev** [comment]
> So, as long as we want to add a new node, we should still keep tracking it in the semantic analyzer. Then, inside the noDeclare rule, we ignore all those bindings that belong to this new node.

OK! I will try to reimplement through this policy you mentioned.

**Pull request closed without merging**